ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveAggregateIncrementalRewritingRule.java

```diff
@@ -99,31 +100,61 @@ public void onMatch(RelOptRuleCall call) {
+    // - true means exist
+    // - null means not exists
+    // Project also needed to encapsulate the view scan by a subquery -> this is required by
+    // CalcitePlanner.fixUpASTAggregateIncrementalRebuild
+    List<RexNode> mvCols = new ArrayList<>(joinLeftInput.getRowType().getFieldCount());
+    for (int i = 0; i < joinLeftInput.getRowType().getFieldCount(); ++i) {
+      mvCols.add(rexBuilder.makeInputRef(
+              joinLeftInput.getRowType().getFieldList().get(i).getType(), i));
+    }
+    mvCols.add(rexBuilder.makeLiteral(true));
+
+    joinLeftInput = call.builder()
+            .push(joinLeftInput)
+            .project(mvCols)
+            .build();
+
+    // 3) Build conditions for join and start adding
     // expressions for project operator
     List<RexNode> projExprs = new ArrayList<>();
     List<RexNode> joinConjs = new ArrayList<>();
-    List<RexNode> filterConjs = new ArrayList<>();
     int groupCount = agg.getGroupCount();
     int totalCount = agg.getGroupCount() + agg.getAggCallList().size();
-    for (int leftPos = 0, rightPos = totalCount;
+    for (int leftPos = 0, rightPos = totalCount + 1;
          leftPos < groupCount; leftPos++, rightPos++) {
       RexNode leftRef = rexBuilder.makeInputRef(
           joinLeftInput.getRowType().getFieldList().get(leftPos).getType(), leftPos);
       RexNode rightRef = rexBuilder.makeInputRef(
           joinRightInput.getRowType().getFieldList().get(leftPos).getType(), rightPos);
       projExprs.add(rightRef);
-      joinConjs.add(rexBuilder.makeCall(SqlStdOperatorTable.EQUALS,
-          ImmutableList.of(leftRef, rightRef)));
-      filterConjs.add(rexBuilder.makeCall(SqlStdOperatorTable.IS_NULL,
-          ImmutableList.of(leftRef)));
+
+      RexNode nsEqExpr = rexBuilder.makeCall(SqlStdOperatorTable.IS_NOT_DISTINCT_FROM,
+              ImmutableList.of(leftRef, rightRef));
+      joinConjs.add(nsEqExpr);
     }
-    // 3) Add the expressions that correspond to the aggregation
+
+    // 4) Create join node
+    RexNode joinCond = RexUtil.composeConjunction(rexBuilder, joinConjs);
+    RelNode join = call.builder()
+            .push(joinLeftInput)
+            .push(joinRightInput)
+            .join(JoinRelType.RIGHT, joinCond)
+            .build();
+
+    int flagIndex = joinLeftInput.getRowType().getFieldCount() - 1;
+    RexNode flagNode = rexBuilder.makeInputRef(
+            join.getRowType().getFieldList().get(flagIndex).getType(), flagIndex);
+
+    // 5) Add the expressions that correspond to the aggregation
     // functions
-    RexNode caseFilterCond = RexUtil.composeConjunction(rexBuilder, filterConjs);
-    for (int i = 0, leftPos = groupCount, rightPos = totalCount + groupCount;
+    for (int i = 0, leftPos = groupCount, rightPos = totalCount + 1 + groupCount;
          leftPos < totalCount; i++, leftPos++, rightPos++) {
       // case when mv2.deptno IS NULL AND mv2.deptname IS NULL then s else source.s + mv2.s end
```

Comment:
This case statement should rely on the flag too?
```
// case mv2.flag IS NULL then s else source.s + mv2.s end
```

Reply:
Unfortunately relying on the flag only is not enough.
Aggregate functions eliminates null values:
```
insert into t1(a,b) values 
(1, NULL),
(1, 10);
select a, sum(b) from t1 group by a
```
results
```
1, 10
```
However
```
select 10 + NULL
```
is `NULL`.

The comment is not valid any more. I will update it.

Reply:
Got it!